File: swift/Sources/benchmarks/Binarytrees.swift
import Foundation

final class BinarytreesObj: BenchmarkProtocol {
  private var n: Int64 = 0
  private var resultVal: UInt32 = 0

  init() {
    n = configValue("depth") ?? 0
  }

  final class TreeNode {
    let item: Int
    let left: TreeNode?
    let right: TreeNode?

    init(item: Int, depth: Int) {
      self.item = item
      if depth > 0 {
        let shift = 1 << (depth - 1)
        left = TreeNode(item: item - shift, depth: depth - 1)
        right = TreeNode(item: item + shift, depth: depth - 1)
      } else {
        left = nil
        right = nil
      }
    }

    func sum() -> UInt32 {
      var total = UInt32(bitPattern: Int32(item)) &+ 1
      if let left = left {
        total &+= left.sum()
      }
      if let right = right {
        total &+= right.sum()
      }
      return total
    }
  }

  func run(iterationId: Int) {
    let root = TreeNode(item: 0, depth: Int(n))
    resultVal &+= root.sum()
  }

  var checksum: UInt32 {
    return resultVal
  }

  func prepare() {}
}

final class BinarytreesArena: BenchmarkProtocol {
  private var n: Int64 = 0
  private var resultVal: UInt32 = 0

  init() {
    n = configValue("depth") ?? 0
  }

  struct TreeNode {
    let item: Int
    var left: Int = -1
    var right: Int = -1
  }

  class TreeArena {
    private var nodes: [TreeNode] = []

    func build(item: Int, depth: Int) -> Int {
      let idx = nodes.count
      nodes.append(TreeNode(item: item))

      if depth > 0 {
        let shift = 1 << (depth - 1)
        let leftIdx = build(item: item - shift, depth: depth - 1)
        let rightIdx = build(item: item + shift, depth: depth - 1)
        nodes[idx].left = leftIdx
        nodes[idx].right = rightIdx
      }

      return idx
    }

    func sum(idx: Int) -> UInt32 {
      let node = nodes[idx]
      var total = UInt32(bitPattern: Int32(node.item)) &+ 1

      if node.left >= 0 {
        total &+= sum(idx: node.left)
      }
      if node.right >= 0 {
        total &+= sum(idx: node.right)
      }

      return total
    }
  }

  func run(iterationId: Int) {
    var arena = TreeArena()
    let rootIdx = arena.build(item: 0, depth: Int(n))
    resultVal &+= arena.sum(idx: rootIdx)
  }

  var checksum: UInt32 {
    return resultVal
  }

  func prepare() {}
}

File: zig/src/binarytrees.zig
const std = @import("std");
const Benchmark = @import("benchmark.zig").Benchmark;
const Helper = @import("helper.zig").Helper;

pub const BinarytreesObj = struct {
    allocator: std.mem.Allocator,
    helper: *Helper,
    n: i64,
    result_val: u32,

    const TreeNode = struct {
        item: i32,
        left: ?*TreeNode,
        right: ?*TreeNode,

        fn create(allocator: std.mem.Allocator, item: i32, depth: i32) !*TreeNode {
            const node = try allocator.create(TreeNode);
            node.* = TreeNode{
                .item = item,
                .left = null,
                .right = null,
            };

            if (depth > 0) {
                node.left = try TreeNode.create(allocator, item - (@as(i32, 1) << @intCast(depth - 1)), depth - 1);
                node.right = try TreeNode.create(allocator, item + (@as(i32, 1) << @intCast(depth - 1)), depth - 1);
            }

            return node;
        }

        fn destroy(self: *TreeNode, allocator: std.mem.Allocator) void {
            if (self.left) |left| left.destroy(allocator);
            if (self.right) |right| right.destroy(allocator);
            allocator.destroy(self);
        }

        fn sum(self: *const TreeNode) u32 {
            var total: u32 = @as(u32, @bitCast(self.item)) +% 1;
            if (self.left) |left| total +%= left.sum();
            if (self.right) |right| total +%= right.sum();
            return total;
        }
    };

    const vtable = Benchmark.VTable{
        .run = runImpl,
        .checksum = resultImpl,
        .deinit = deinitImpl,
    };

    pub fn init(allocator: std.mem.Allocator, helper: *Helper) !*BinarytreesObj {
        const n = helper.config_i64("Binarytrees::Obj", "depth");

        const self = try allocator.create(BinarytreesObj);
        errdefer allocator.destroy(self);

        self.* = BinarytreesObj{
            .allocator = allocator,
            .helper = helper,
            .n = n,
            .result_val = 0,
        };
        return self;
    }

    pub fn deinit(self: *BinarytreesObj) void {
        self.allocator.destroy(self);
    }

    pub fn asBenchmark(self: *BinarytreesObj) Benchmark {
        return Benchmark.init(self, &vtable, self.helper, "Binarytrees::Obj");
    }

    fn runImpl(ptr: *anyopaque, iteration_id: i64) void {
        _ = iteration_id;
        const self: *BinarytreesObj = @ptrCast(@alignCast(ptr));

        const root = TreeNode.create(self.allocator, 0, @intCast(self.n)) catch return;
        defer root.destroy(self.allocator);

        self.result_val +%= root.sum();
    }

    fn resultImpl(ptr: *anyopaque) u32 {
        const self: *BinarytreesObj = @ptrCast(@alignCast(ptr));
        return self.result_val;
    }

    fn deinitImpl(ptr: *anyopaque) void {
        const self: *BinarytreesObj = @ptrCast(@alignCast(ptr));
        self.deinit();
    }
};

pub const BinarytreesArena = struct {
    allocator: std.mem.Allocator,
    helper: *Helper,
    n: i64,
    result_val: u32,

    const TreeNode = struct {
        item: i32,
        left: ?*TreeNode,
        right: ?*TreeNode,

        fn create(allocator: std.mem.Allocator, item: i32, depth: i32) !*TreeNode {
            const node = try allocator.create(TreeNode);
            node.* = TreeNode{
                .item = item,
                .left = null,
                .right = null,
            };

            if (depth > 0) {
                node.left = try TreeNode.create(allocator, item - (@as(i32, 1) << @intCast(depth - 1)), depth - 1);
                node.right = try TreeNode.create(allocator, item + (@as(i32, 1) << @intCast(depth - 1)), depth - 1);
            }

            return node;
        }

        fn sum(self: *const TreeNode) u32 {
            var total: u32 = @as(u32, @bitCast(self.item)) +% 1;
            if (self.left) |left| total +%= left.sum();
            if (self.right) |right| total +%= right.sum();
            return total;
        }
    };

    const vtable = Benchmark.VTable{
        .run = runImpl,
        .checksum = resultImpl,
        .deinit = deinitImpl,
    };

    pub fn init(allocator: std.mem.Allocator, helper: *Helper) !*BinarytreesArena {
        const n = helper.config_i64("Binarytrees::Arena", "depth");

        const self = try allocator.create(BinarytreesArena);
        errdefer allocator.destroy(self);

        self.* = BinarytreesArena{
            .allocator = allocator,
            .helper = helper,
            .n = n,
            .result_val = 0,
        };
        return self;
    }

    pub fn deinit(self: *BinarytreesArena) void {
        self.allocator.destroy(self);
    }

    pub fn asBenchmark(self: *BinarytreesArena) Benchmark {
        return Benchmark.init(self, &vtable, self.helper, "Binarytrees::Arena");
    }

    fn runImpl(ptr: *anyopaque, iteration_id: i64) void {
        _ = iteration_id;
        const self: *BinarytreesArena = @ptrCast(@alignCast(ptr));

        var arena = std.heap.ArenaAllocator.init(self.allocator);
        defer arena.deinit();

        const allocator = arena.allocator();

        const root = TreeNode.create(allocator, 0, @intCast(self.n)) catch return;
        self.result_val +%= root.sum();
    }

    fn resultImpl(ptr: *anyopaque) u32 {
        const self: *BinarytreesArena = @ptrCast(@alignCast(ptr));
        return self.result_val;
    }

    fn deinitImpl(ptr: *anyopaque) void {
        const self: *BinarytreesArena = @ptrCast(@alignCast(ptr));
        self.deinit();
    }
};
